pkg/issuer/acme/dns/util/wait.go

```diff
@@ -168,6 +168,87 @@ func dnsQuery(fqdn string, rtype uint16, nameservers []string, recursive bool) (
 	return
 }
 
+func ValidateCAA(domain string, issuerID []string, iswildcard bool) error {
+	// see https://tools.ietf.org/html/rfc6844#section-4
+	// for more information about how CAA lookup is performed
+	fqdn := ToFqdn(domain)
+
+	issuerSet := make(map[string]bool)
+	for _, s := range issuerID {
+		issuerSet[s] = true
+	}
+
+	var caas []*dns.CAA
+	for {
+		// follow at most 8 cnames per label
+		queryDomain := fqdn
+		var msg *dns.Msg
+		var err error
+		for i := 0; i < 8; i++ {
+			//TODO(dmo): figure out if we need these servers to be configurable as well
+			msg, err = dnsQuery(queryDomain, dns.TypeCAA, RecursiveNameservers, true)
+			if err != nil {
```

Comment:
If a zone does not have a CAA record set, will it return SERVFAIL and cause issues?

Reply:
Do we use an old version of the miekg DNS library that turned certain "successful" DNS queries into errors? otherwise this error should only happen on actual communications failures

Reply:
If a zone returns `SERVFAIL` (or `NOTIMPL`) you would want to catch that in this precheck. Let's Encrypt would reject either case as an invalid CAA policy: https://letsencrypt.org/docs/caa/#servfail